### DISCUSSION TITLE
Contributing translations

### DISCUSSION THREAD

**Discussion by itagagaki** [1Remote/1Remote#756] [_Development]
I am thinking of contributing Japanese.
You say to edit glossary.csv, not any .xaml. However, ja-jp.xaml and glossary.csv in repository do not match. The Japanese columns in glossary.csv are all blank.
What should I do?

I have another suggestion.
```
var about = new System.Windows.Forms.ToolStripMenuItem(IoC.Translate("About") + $" {Assert.APP_DISPLAY_NAME}");
```
This line in TaskTrayServices.cs is fine for languages that are represented in this word order, but not for other languages (e.g. Japanese).
So the only way to do this is to make sure that `Translate(‘About’)` represents everything.
For example, Japanese resource might look like this.
```
<s:String x:Key="About">1Remoteについて</s:String>
```

**Comment by VShawn** [comment]
> You say to edit glossary.csv, not any .xaml. However, ja-jp.xaml and glossary.csv in repository do not match. The Japanese columns in glossary.csv are all blank.

The `glossary.csv` file stores translations that have been reviewed by humans.

Blank fields mean there was no translator have contributed translations yet. The blanks in `glossary.csv` will be translated to corresponding languages by Google Translate when running the conver_glossary_to_xaml.bat script. The script generating `glossary_translated_by_google.csv` file, then `glossary_translated_by_google.csv` file is converted into xaml resources.

So.All You need to do is fill in the blanks and then run `conver_glossary_to_xaml.bat` to replace the current machine-translated Japanese text with your translations.

> I have another suggestion.

Thank you suggestion. This was my oversight, as I didn't consider languages with inverted word order like Japanese.
To simplify the work, I suggest reducing `IoC.Translate("About") + $" {Assert.APP_DISPLAY_NAME}"` to `IoC.Translate("About")`. And If there are other text with similar situations, please feel free to point them out for further discussion :)

**Comment by itagagaki** [comment]
Okay. I understand!

**Comment by itagagaki** [comment]
I will continue to write here about issues that arose while working on the translation.
First, I would like to clarify the meaning of the following texts.

> We can not insert into database:

   It is unclear what to insert and where to insert it.
   Inserting data into a table?

> We can not update on database:

   Updating a table?

> We can not read from database:

   Couldn't read a table from the database?
   Or couldn't get something from a table?

> We can not select from database:

   I understand that this message is for when data cannot be extracted (selected) from a table, is this correct?

> We can not delete from database:

   Couldn't delete a table from the database?
   Or couldn't delete something from a table?

**Reply by VShawn** [reply]
Ah, I didn't see this one.

![image](https://github.com/user-attachments/assets/1ad4696b-a81f-4e77-8188-ecef56bd7340)

Since user may have multiple database setting in their app, these prompts are to tell user which database has failed. not to indicate which table has failed.

note: these sentences are followed by colons, and error messages are generated in the code explaining which database it cannot insert into and what error occurred. Because db faults are rare, this is my `lazy way? or lazy development?` to give users limited error indications.

E.G.

> `We can not insert into database:`

https://github.com/1Remote/1Remote/blob/2d18bbf07c95a5e533d01012d1ba837db8d7f155/Ui/Service/DataSource/DAO/Dapper/DapperDataBase.cs#L229-L252

Once an error occurs, the error message will be assembled here:

https://github.com/1Remote/1Remote/blob/2d18bbf07c95a5e533d01012d1ba837db8d7f155/Ui/Service/DataSource/DAO/IDataBase.cs#L39-L44

Demo of the error message：

```
We can not insert into database: AI
beacuse:
connection failed (This specific error reason is provided by the database component: dapper)
```

**Reply by itagagaki** [reply]
Hmmm, I have a feeling this will affect the English translations as well.
My understand is that the object term of one sentence is the data and the object of another sentence the the database.
So I think we can say the following?

We can not connect database --> Couldn't connect to database
We can not create tables on database --> Couldn't create tables on database (or Couldn't add tables to database)
We can not insert into database --> Couldn't insert data into database
We can not update on database --> Couldn't update data in database
We can not read from database --> Couldn't read database (or Couldn't read data from database? I'm not sure...)
We can not select from database --> Couldn't extract data from the database
We can not delete from database --> Couldn't delete data from database

Is this correct? I am a bit unsure.

**Reply by VShawn** [reply]
I don't mind, I think the translation here should just convey the meaning. Technicians should be able to understand the meaning.

**Reply by itagagaki** [reply]
Okay, if the English phrases on the right side of the above comment are semantically correct, then I will translate them into Japanese with their meanings. Okay?

**Reply by VShawn** [reply]
sure thing

**Comment by itagagaki** [comment]
I can't figure out how to see the following terms/phrases in the app.

- Lock (row 31)
- Selected (row 66)
- host_reconecting_info (row 93)
- system_options_data_security_info_data_processing (row 106)
- system_options_data_security_error_can_not_open (row 107)
- string_database_error_permission_denied (row 108)
- server_editor_different_options (row 122)
- server_editor_tags_tag (row 132)
- Run bat before connect (row 136)
- Run bat after disconnected (row 137)
- Open SFTP (row 141)
- Hi there, (row 240)
- Do not show this again in current version (row 245)
- Please select a Data Source (row 256)
- Pause (row 283)
- Updating (row 284)
- Reconnecting (row 285)
- Next update check (row 286)
- Next auto reconnect (row 287)
- Please complete the windows credentials verification (row 294)
- Do you want to replace the current parameter list with preset value? (row 310)
- Remember the information above (317)

Are the following no longer in use?

- Upload (row 71)
- Automatically focus on the content when selecting a tab (row 313)

And I think the following are not in use.

- server_editor_gateway_bypass_gateway_for_local_address (row 192)
- Show environment variables (row 234)
- XXX is not existed! (row 271)

Please also respond to my previous https://github.com/1Remote/1Remote/discussions/756#discussioncomment-11363869

**Reply by VShawn** [reply]
uh. maybe by searching? 

like system_options_data_security_info_data_processing  

https://github.com/1Remote/1Remote/blob/2d18bbf07c95a5e533d01012d1ba837db8d7f155/Ui/View/ServerList/ServerListPageViewModel.cs#L604


> Are the following no longer in use?

After many refactorings, some of the translated text may be useless, and I should write a filtering python script to pick them out.


> XXX is not existed!

it do still in use

https://github.com/1Remote/1Remote/blob/2d18bbf07c95a5e533d01012d1ba837db8d7f155/Ui/Model/TagActionHelper.cs#L89-L91

**Comment by itagagaki** [comment]
Thanks for the answer.
Sorry, but I would also appreciate an answer to my database-related question.

> uh. maybe by searching?

Excuse me, I would like to know the procedure by which I can see these texts actually displayed by the application in the window.
The reason for this is that the appropriate translation may vary depending on the context.

> > XXX is not existed!
> 
> it do still in use

It's not "XXX is not existed!" but "XXX is already existed!".

**Reply by VShawn** [reply]
> It's not "XXX is not existed!" but "XXX is already existed!".

Sorry, my mistake... my mind is messed up because of a fever

>  I would like to know the procedure by which I can see these texts actually displayed by the application in the window.

But some of them cannot be found directly in the APP

## let me show you

### seems unused:

~Lock (row 31)~
~Selected (row 66)~ 
~server_editor_tags_tag (row 132)~

### 

> `host_reconecting_info (row 93) `

Displayed during RDP reconnection, I can not get a screens hoot for you 

> `system_options_data_security_info_data_processing (row 106)`

Displayed during importing data from json

> system_options_data_security_error_can_not_open (row 107)

![image](https://github.com/user-attachments/assets/3ff765db-efba-4684-b51e-14eccde9b25d)

> string_database_error_permission_denied (row 108)

when your database is not writable

![image](https://github.com/user-attachments/assets/2521344f-2e9c-48c2-8e70-fb8a1cecc730)

> server_editor_different_options (row 122)

bulk edit

![image](https://github.com/user-attachments/assets/c784e1dd-2c98-4db8-9c30-0ef9798bb449)

> Run bat before connect (row 136)
> Run bat after disconnected (row 137) 

hint text

![image](https://github.com/user-attachments/assets/7f340b57-c8f2-41b8-8fc6-c19a2d303dbf)


> Open SFTP (row 141)

![image](https://github.com/user-attachments/assets/36604f4c-914f-4dff-82e4-51380a449ef4)


> Hi there, (row 240)
> Do not show this again in current version (row 245)

![image](https://github.com/user-attachments/assets/ee17b729-65c5-4495-b03c-801eb6c42217)

> Please select a Data Source (row 256)

When you have to database and decide to add a server session

![image](https://github.com/user-attachments/assets/001e61fb-c017-42af-8586-353da5cefd9c)

> Pause (row 283)
> Updating (row 284)
> Next update check (row 286)


indicate that the datasource is stopping updating

![image](https://github.com/user-attachments/assets/953723dc-e35f-47b1-9799-2ca0f8da4339)

normally it should display a countdown and reload the data from the database when the time is up.

![image](https://github.com/user-attachments/assets/e928f18e-8e50-45de-b029-12c5ef53f73d)

when it reading from the database the `Updating` shown

![image](https://github.com/user-attachments/assets/f7031ec5-c3b1-4594-827b-164cc497964a)

> Reconnecting (row 285)
> Next auto reconnect (row 287)

when the database is lost connection and trying to reconnect: https://1remote.org/usage/database/use-online-database/#use-mysql

similar with `Updating` & `Next update check`, but I can not get a screens shoot currently

> Please complete the windows credentials verification (row 294)

when you are canceling the Windows hello feature

![image](https://github.com/user-attachments/assets/96be0509-c5d5-4cf6-be14-f108fd672b02)

![image](https://github.com/user-attachments/assets/6008895d-d985-4a1e-9d71-f8c78f3a20d6)

> Do you want to replace the current parameter list with preset value? (row 310)
Remember the information above (317)

when setting up app protocol

![image](https://github.com/user-attachments/assets/37648274-a2a7-428f-936f-953de331cc41)

![image](https://github.com/user-attachments/assets/b93e57f8-f7ba-47dc-8fb2-5e98eb6f6397)

**Reply by itagagaki** [reply]
Thank you! I will try to add the corresponding Japanese translation.

**Reply by itagagaki** [reply]
Sorry, I could not get the "Do you want to replace the current parameter list with the preset value?" window to open.

In LocalAppFormViewModel.cs

```
        public string ExePath
        {
            get => New.ExePath;
            set
            {
                if (New.ExePath != value)
                {
                    New.ExePath = value;
                    RaisePropertyChanged();

                    var t = AppArgumentHelper.GetPresetArgumentList(ExePath);
                    if (t != null)
                    {
```
t is always null even if CMD parameter is set.

**Reply by VShawn** [reply]
in app protocol
1. click `select exe` button, select putty.exe, the `paramter list` will be filled with putty.exe's cli paramaters.
2. 1. click `select exe` button again, selecte winscp.exe, popup shows, the `paramter list` will be replaced by winscp.exe 's cli paramaters if `Yes` button in the popup is clcicked.

**Comment by itagagaki** [comment]
UPDATE: Japanese localization is complete on my end, except for the following
31, 66, 71, 132, 192, 234, 271, 313 - seems unused
310 - due to #787
